README.md
DynPick ROS driver
==================================

![WDF-6M200-3](http://www.wacoh-tech.com/img/img_model_wdf_6m200_3.jpg "WDF-6M200-3")

ROS driver for [Wacoh-tech force sensor](http://www.wacoh-tech.com/en/products/dynpick/).

Prerequisite
==============

 * (Only tested on) Ubuntu 12.04 and 14.04 64bit.
 * ROS Hydro or Indigo needs to be installed.

Install
==========

 1. Create `/etc/udev/rules.d/80-wakoh.rules` with the following line:

 ```
## see https://github.com/tork-a/dynpick_driver
ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="ea60", MODE="0666"
 ```

 2. Run `$ sudo service udev restart`

 3. Plug the sensor USB to the Ubuntu computer. Also make sure that the following device was detected.

 ```
$ lsusb 
:
Bus 003 Device 004: ID 10c4:ea60 Cygnal Integrated Products, Inc. CP210x Composite Device
 ```

 4. Obtain `dynpick_driver` package.

 4-a. Via DEB (RECOMMENDED). Run `sudo apt-get install ros-hydro-dynpick-driver`.

 4-b. (Source build. Not recommended) clone this repository and build.

   ```
$ cd %YOUR_CATKING_WORKSPACE%/src
$ git clone https://github.com/tork-a/dynpick_driver.git
$ cd %YOUR_CATKING_WORKSPACE%
$ catkin_make
   ```

Operation
==========

You can view the force and torque vectors in rviz by launching sample.launch:

```
 $ roslaunch dynpick_driver sample.launch
```

 You can configure some things by passing arguments:

 ```
 $ roslaunch dynpick_driver sample.launch device:=/dev/ttyUSB0 rate:=500 sensor_frame_id:=/sensor rvizconfig:=`rospack find dynpick_driver`/launch/sample.rviz
```

 where:

 * `device`: port name of the connected sensor device
 * `rate`: refresh rate
 * `sensor_frame_id`: This name will be used as a [tf](http://wiki.ros.org/tf) frame of the connected sensor device
 * `rvizconfig`: name of the [RViz configuration file](http://wiki.ros.org/rviz/UserGuide#Configurations) that can be used for storing your own `RViz` setting
 * `topic`: name of the published topic (WrenchStamped). Default is '/force'

Video is posted on ROS wiki (http://wiki.ros.org/action/subscribe/dynpick_driver)

[![Build Status](https://travis-ci.org/tork-a/dynpick_driver.png)](https://travis-ci.org/tork-a/dynpick_driver)
--------------------------------------------------------------------------------------------------------------------------------------------

Document
========

## ROS API

### Published Topics

`/force` ([geometry_msgs/WrenchStamped](http://docs.ros.org/api/geometry_msgs/html/msg/WrenchStamped.html))

### Services

/dynpick_driver_node/get_loggers ([roscpp/GetLoggers](http://wiki.ros.org/roscpp))
/dynpick_driver_node/set_logger_level ([roscpp/SetLoggerLevel](http://wiki.ros.org/roscpp))

### Parameters

`/dynpick_driver_node/device`

`/dynpick_driver_node/frame_id`

`/dynpick_driver_node/rate`

Other
--------

[API doc is available](http://docs.ros.org/hydro/api/dynpick_driver/html/annotated.html) also on ROS domain.

DEB build status
================

| ROS Distro   | Source deb | Development Branch (travis)  | Development branch (ros.org) | Release Branch | binarydeb Precise AMD64 | Documentation (ros.org) |
| ------------- | ------------- | ---------------------------- | ---------------------------- | -------------- | ----------------------- | ----------------------- |
| Indigo | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=ros-indigo-dynpick_driver_sourcedeb)](http://jenkins.ros.org/job/ros-indigo-dynpick_driver_sourcedeb/) | [![Build Status](https://travis-ci.org/tork-a/dynpick_driver.png?branch=indigo-devel)](https://travis-ci.org/tork-a/dynpick_driver) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=devel-indigo-dynpick_driver)](http://jenkins.ros.org/job/devel-indigo-dynpick_driver/) | [![Build Status](https://travis-ci.org/lagadic/dynpick_driver.png?branch=indigo)](https://travis-ci.org/lagadic/dynpick_driver) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=ros-indigo-dynpick_driver_binarydeb_precise_amd64)](http://jenkins.ros.org/job/ros-indigo-dynpick_driver_binarydeb_precise_amd64/) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=doc-indigo-dynpick_driver)](http://jenkins.ros.org/job/doc-indigo-dynpick_driver/) |
| Hydro | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=ros-hydro-dynpick_driver_sourcedeb)](http://jenkins.ros.org/job/ros-hydro-dynpick_driver_sourcedeb/) | [![Build Status](https://travis-ci.org/tork-a/dynpick_driver.png?branch=hydro-devel)](https://travis-ci.org/tork-a/dynpick_driver) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=devel-hydro-dynpick_driver)](http://jenkins.ros.org/job/devel-hydro-dynpick_driver/) | [![Build Status](https://travis-ci.org/lagadic/dynpick_driver.png?branch=hydro)](https://travis-ci.org/lagadic/dynpick_driver) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=ros-hydro-dynpick_driver_binarydeb_precise_amd64)](http://jenkins.ros.org/job/ros-hydro-dynpick_driver_binarydeb_precise_amd64/) | [![Build Status](http://jenkins.ros.org/buildStatus/icon?job=doc-hydro-dynpick_driver)](http://jenkins.ros.org/job/doc-hydro-dynpick_driver/) |

[Devel Test Status](http://wiki.ros.org/regression_tests#Development_Tests)
-------------------------------------------------------------------------------------

[![Hydro Test Satus](http://jenkins.ros.org/job/devel-hydro-dynpick_driver/test/trend?job)](http://jenkins.ros.org/job/devel-hydro-dynpick_driver/)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package dynpick_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Fix ROS buildfarm error

  * revert `#27 <https://github.com/tork-a/dynpick_driver/issues/27>`_  (`#40 <https://github.com/tork-a/dynpick_driver/issues/40>`_ )
  * add header file and use `fputs` instaed of `fprintf` to avoid compile error
     * warning: implicit declaration of function
     * warning: format not a string literal and no format arguments

* [sample.launch] args improvement (`#38 <https://github.com/tork-a/dynpick_driver/issues/38>`_)

  * [sample.launch] Choose running RViz or not.
  * [sample.launch] pass wrench topic name.

* Adding more build-in features by sensor (`#39 <https://github.com/tork-a/dynpick_driver/issues/39>`_)

  * Fix reading problems with A6200 version
    the reply seems to differ for some reason, however this fix it by cleaning up the socket after sensitivity request and filter request
  * Set built-in filter at startup
  * Add automatical adjustment of LSB/N

    * generalized "readFromSocket" function
    * flushing socket at beginning (to avoid byte offset)
    * receiving calibration (LSB/N and LSB/Nm) from sensor and multiplying it to data
    * tabs vs. spaces cleanup

* Contributors: Isaac I.Y. Saito, Kei Okada, Lorenz Halt

0.1.1 (2016-12-20)
------------------
* [fix] Add a missing folder to be installed `#37 <https://github.com/tork-a/dynpick_driver/issues/37>`_
* [capability] Add a tested product model. `#35 <https://github.com/tork-a/dynpick_driver/issues/35>`_
* Contributors: Isaac I.Y. Saito

0.1.0 (2016-09-15)
------------------
* ADD urdf model `#32 <https://github.com/tork-a/dynpick_driver/issues/32>`_
  * Changed sample launch to show model instead of static tf-frame
* Add tare service `#30 <https://github.com/tork-a/dynpick_driver/issues/30>`_
  * uses std::srvs::Trigger
  * ordered dependencies alphabetically
  * Need to send offset_reset command several times (3 at my tests) to work
* Contributors: Lorenz Halt

0.0.11 (2016-06-02)
-------------------
* Workaround for ROS buildfarm error. `#27 <https://github.com/tork-a/dynpick_driver/issues/27>`_
* Contributors: Isaac I.Y. Saito

0.0.10 (2016-06-02)
-------------------
* [feat] Add test binary for measuring device throughput on Ubuntu (tested against QNX). `#25 <https://github.com/tork-a/dynpick_driver/issues/25>`_
* [doc] Add a tested product ID. `#26 <https://github.com/tork-a/dynpick_driver/issues/26>`_
* Contributors: Isaac I.Y. Saito

0.0.9 (2016-03-23)
------------------
* [fix] run.launch not publishing Wrench topic issue.
* [feat] Rename run.launch --> driver.launch
* [sys] add test
* [sys] add travis.yml
* [doc] Switched to sphinx; this doxygen setting seems not working, and the doc we want by rosdoc is not about api docs but rst files. (`#20 <https://github.com/tork-a/dynpick_driver/issues/20>`_)
* Contributors: Kei Okada, Isaac I.Y. Saito

0.0.8 (2015-08-14)
------------------
* [doc] Utilize rosdoc_lite.
* Contributors: Isaac I.Y. Saito

0.0.7 (2015-08-12)
------------------
* [feat] Whether or not running RViz
* Contributors: Isaac I.Y. Saito

0.0.6 (2015-08-11)
------------------
* [fix] Missing launch file variables
* [doc] Add more doc
* Contributors: Ron Tajima, Isaac I.Y. Saito

0.0.5 (2015-04-01)
------------------
* (Feature) Do while to read enough data (fix for `#1 <https://github.com/tork-a/dynpick_driver/issues/1>`_)
* (Doc) Add to readme new launch usage.
* Contributors: Kei Okada, Isaac I.Y. Saito

0.0.4 (2015-03-23)
------------------
* Reusable launch files.
* Contributors: Isaac I.Y. Saito

0.0.3 (2014-12-29)
------------------
* Fix/Typo, this may break downstream packages `#7 <https://github.com/tork-a/dynpick_driver/issues/7>`_ from k-okada/fix_typo
* Fix/Install missing resource. `#6 <https://github.com/tork-a/dynpick_driver/issues/6>`_ from 130s/fix/install_launch
* Contributors: Kei Okada, Isaac I.Y. Saito

0.0.2 (2014-12-24)
------------------
* Initial commit
* Add doc
* Contributors: Kei Okada, Isaac I.Y. Saito


src/test-com.c
// test-com.c

/*
 * Build with `gcc -o`. See https://github.com/start-jsk/rtmros_hironx/blob/6eb507f3ad0ed368d909960dd59ddaac0f77f496/hironx_ros_bridge/robot/dynpick/Makefile#L9
 */

#define 	DEBUGSS	0

#include	<stdio.h>
#include	<fcntl.h>
#include	<errno.h>
#include	<time.h>
#include	<termios.h>
#include	<string.h>
#include <unistd.h>
#include <math.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/select.h>
#if defined(__QNX__)
  #include <sys/slog.h>
#endif
#include <sys/time.h>

#define true		1
#define false		0

int kbhit(void) {

	struct termios oldt, newt;
	int ch;
	int oldf;

	tcgetattr(STDIN_FILENO, &oldt);
	newt = oldt;
	newt.c_lflag &= ~(ICANON | ECHO);
	tcsetattr(STDIN_FILENO, TCSANOW, &newt);
	oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
	fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

	ch = getchar();

	tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
	fcntl(STDIN_FILENO, F_SETFL, oldf);

	if (ch != EOF) {

		ungetc(ch, stdin);
		return 1;
	}

	return 0;
}

#define cfsetspeed(term, baudrate)\
   cfsetispeed(term, baudrate);\
   cfsetospeed(term, baudrate);

int com_write(int fd, const char* msg) {
	// IO is currently non-blocking. This is what we want for the more common read case.
	int origflags = fcntl(fd, F_GETFL, 0);
	fcntl(fd, F_SETFL, origflags & ~O_NONBLOCK); // @todo can we make this all work in non-blocking?
	ssize_t len = strlen(msg);
	ssize_t retval = write(fd, msg, len);
	int fputserrno = errno;
	fcntl(fd, F_SETFL, origflags | O_NONBLOCK);
	errno = fputserrno; // Don't want to see the fcntl errno below.

	if (retval != -1) {
		return retval;
	}
	printf("fputs fialed %d: %s", errno, strerror(errno));
	return -1;
}

int main() {
	int status;
	//FILE			*fd;
	int fd;
	char fname[64];
	char devname_in[64];
	char devname_out[64];
	char str[256];
	unsigned short data[6];
	int comNo;
	int tick;
	int clk, clkb, clkb2, clk0;
	int tw;
	int num;
	int n, len;

	//fd = NULL;
	fd = -1;

	start:
	// Open COM port
	printf("Enter COM port > ");
	scanf("%d", &comNo);
	comNo = 0;
	sprintf(devname_out, "/dev/ttyUSB%d", comNo);
	#if defined(__QNX__)
	    comNo = 1;
	    sprintf(devname_out, "/dev/serusb%d", comNo);
	#endif
	printf("Open %s for output\n", devname_out);
	fd = open(devname_out, O_RDWR | O_NOCTTY | O_NONBLOCK);

	if (fd < 0 || fd < 0)
		goto over;

	// Get Sampling Frequency
	tw = 16;
	printf("Enter sampling time (ms) > ");
	fflush(stdout);
	//scanf("%d", &tw);
	printf("Sampling time = %d ms\n", tw);

	printf("Enter File name > ");
	fflush(stdout);
	//scanf("%s", fname);
	//fd = fopen(fname, "w");
	//fd = fopen("test.txt", "w");
	//if (!fd)
	//	goto over;
	#if defined(__QNX__)
        slogf(0, _SLOG_INFO, "Started  input: %s fd = %d\n", devname_in, fd);
	    slogf(0, _SLOG_INFO, "        output: %s fd = %d\n", devname_out, fd);
	#endif

	// Set Commport Settings
	SetComAttr(fd);

	// Read Data
	printf("=== record data ===\n");
	clk0 = clock() / (CLOCKS_PER_SEC / 1000);
	clkb = 0;
	clkb2 = 0;
	num = 0;

	// Data Requests
	struct timeval wr0, wr1, loop0, loop1, fscan0, fscan1;
	gettimeofday(&wr0, NULL);
	n = write(fd, "R", 1);
	gettimeofday(&wr1, NULL);
	// tcdrain(fd); // Old legacy that is no longer needed. Besideds that, on QNX6.5.0 this is found taking unnecessarily too long (say 100msec).
#define DELTA_SEC(start, end) (end.tv_sec - start.tv_sec + (end.tv_usec - start.tv_usec)/1e6)
	printf("write took %7.3f [msec]\n", DELTA_SEC(wr0, wr1) * 1000);
	printf("write data (ret %d)\n", n);

	while (true) {
		// Wait for Sampling rate
		while (true) {
			clk = clock() / (CLOCKS_PER_SEC / 1000) - clk0;
			if (clk >= clkb + tw) {
				clkb = clk / tw * tw;
				break;
			}
		}

		// Data Request
		gettimeofday(&wr0, NULL);
		n = write(fd, "R", 1);
		gettimeofday(&wr1, NULL);
		// tcdrain(fd);  //  The tcdrain() function waits until all output has been physically transmitted to the device associated with fd, or until a signal is received.
		printf("write took %7.3f [msec]\n", DELTA_SEC(wr0, wr1) * 1000);
		printf("write data (ret %d)\n", n);

		// Get Single data
#define DATA_LENGTH 27
		len = 0;
		bzero(str, 27);  // Setting the first 27 bytes of the area starting `str` to zero.
		while (len < DATA_LENGTH) {
			gettimeofday(&loop0, NULL);
			n = read(fd, str + len, DATA_LENGTH - len);
			printf("read data (ret %d, %d bytes read))\n", n, len + n);
			if (n > 0) {
				len += n;
			} else {
				char *pmesg = strerror(errno);
				fprintf(stderr, "failed to read data (%d): %s (%d)\n", n, pmesg,
						errno);
				usleep(10000);
				//goto loop_exit;
			}
			gettimeofday(&loop1, NULL);
			printf("Loop (read) took %7.3f [msec]\n", DELTA_SEC(loop0, loop1) * 1000);
		}
		//goto skip;
		loop_exit: {
			int i;
			for (i = 0; i < DATA_LENGTH; i++) {
				fprintf(stderr, "%02x:", 0x0000ff & str[i]);
			}
			printf("\n");
		}

		gettimeofday(&fscan0, NULL);
		sscanf(str, "%1d%4hx%4hx%4hx%4hx%4hx%4hx", &tick, &data[0], &data[1],
				&data[2], &data[3], &data[4], &data[5]);

		sprintf(str, "%05d,%d,%05d,%05d,%05d,%05d,%05d,%05d\n", clk / tw * tw,
				tick, data[0], data[1], data[2], data[3], data[4], data[5]);
		gettimeofday(&fscan1, NULL);
		printf("sscanf, sprintf took %7.3f [msec]\n", DELTA_SEC(fscan0, fscan1) * 1000);

		fprintf(stderr, str);
		//fprintf(fd, str);
		num++;

		skip: if (clk >= 10000)
			break;

		// Dsiaply Console
		if (clk >= clkb2 + 1000) {
			printf(str);
			if (kbhit() && getchar() == '.')
				break;
			clkb2 = clk / 1000 * 1000;
		}
	}

	over: if (fd >= 0) {
		close(fd);
		fd = -1;
	}

	printf("=== num = %d ===\n", num);

	printf("exit (y / n) ? > ");
	fflush(stdout);
	exit(0);   //scanf("%s", str);
	if (str[0] == 'y') {
//		exit(0);
	} else {
		goto start;
	}
}

void clear_packet(int fd) {

	// clear existing packet
	int oldf = fcntl(fd, F_GETFL, 0);
	fcntl(fd, F_SETFL, oldf | O_NONBLOCK);
	unsigned char c;
	while (read(fd, &c, 1) != EOF)
		;
	fcntl(fd, F_SETFL, oldf);
}

//#define  B921600  B57600
#define   B921600 0010007

int SetComAttr(int fdc) {
	if (fdc < 0) {
		char *pmesg = strerror(errno);
		fprintf(stderr, "failed to open : %s\n", pmesg);
		goto over;
	}

	struct termios term;
	int res = tcgetattr(fdc, &term);
	if (res < 0) {
		char *pmesg = strerror(errno);
		fprintf(stderr, "failed to tcgetattr(): %s\n", pmesg);
		goto over;
	}
	cfmakeraw(&term);
	res = cfsetspeed(&term, 921600)
	;
//	 res = cfsetspeed(&term, 230400);
	if (res < 0) {

		char *pmesg = strerror(errno);
		fprintf(stderr, "failed to cfsetspeed(): %s\n", pmesg);
		goto over;
	}

	// settings for qnx
	term.c_iflag |= IGNPAR;            // Ignore characters with parity errors
	term.c_cflag |= (CLOCAL | CREAD);  // needed for QNX 6.3.2
	term.c_cflag &= ~PARENB;           // disable parity check
	term.c_cflag |= CS8;               // 8 data bit
	term.c_cflag &= ~CSTOPB;           // 1 stop bit
	//term.c_lflag = IEXTEN;
	term.c_oflag = 0; ///added
	term.c_lflag &= ~(ECHO | ECHOCTL | ECHONL);  // disable ECHO

	//
	term.c_iflag &= ~INPCK;

	// settings for dynpick
	term.c_cc[VINTR] = 0; /* Ctrl-c */
	term.c_cc[VQUIT] = 0; /* Ctrl-? */
	term.c_cc[VERASE] = 0; /* del */
	term.c_cc[VKILL] = 0; /* @ */
	term.c_cc[VEOF] = 4; /* Ctrl-d */
	term.c_cc[VTIME] = 0;
	term.c_cc[VMIN] = 1;
	//term.c_cc[VSWTC]    = 0;     /* '?0' */
	term.c_cc[VSTART] = 0; /* Ctrl-q */
	term.c_cc[VSTOP] = 0; /* Ctrl-s */
	term.c_cc[VSUSP] = 0; /* Ctrl-z */
	term.c_cc[VEOL] = 0; /* '?0' */
	term.c_cc[VREPRINT] = 0; /* Ctrl-r */
	term.c_cc[VDISCARD] = 0; /* Ctrl-u */
	term.c_cc[VWERASE] = 0; /* Ctrl-w */
	term.c_cc[VLNEXT] = 0; /* Ctrl-v */
	term.c_cc[VEOL2] = 0; /* '?0' */
#ifdef __QNX__
	term.c_cflag &= ~(IHFLOW | OHFLOW);
#endif

	// settings for qnx
	term.c_cc[VMIN] = 0;
	term.c_cc[VTIME] = 2;

	res = tcsetattr(fdc, TCSANOW, &term);
	if (res < 0) {
		char *pmesg = strerror(errno);
		fprintf(stderr, "failed to tcsetattr(): %s\n", pmesg);
		goto over;
	}

	over: return (res);
}


src/main.cpp
#include <ros/ros.h>
#include <geometry_msgs/WrenchStamped.h>
#include <std_srvs/Trigger.h>

#include    <stdio.h>
#include    <fcntl.h>
#include    <time.h>
#include    <termios.h>
#include    <string.h>
#include    <unistd.h>

#include <mutex>
#include <condition_variable>

#define true                1
#define false               0

// TODO: Find out why and fix the multiple try approach
#define RESET_COMMAND_TRY   3       // It only works when send several times.

#define DATA_LENGTH 27
#define CALIB_DATA_LENGTH 46

std::mutex m_;
std::condition_variable cv_;
int offset_reset_ = 0;

int SetComAttr(int fdc)
    {
    int n;

    struct termios term;


    // Set baud rate
    n = tcgetattr(fdc, &term);
    if (n < 0)
        goto over;

    bzero(&term, sizeof(term));

    term.c_cflag = B921600 | CS8 | CLOCAL | CREAD;
    term.c_iflag = IGNPAR;
    term.c_oflag = 0;
    term.c_lflag = 0;/*ICANON;*/

    term.c_cc[VINTR]    = 0;     /* Ctrl-c */
    term.c_cc[VQUIT]    = 0;     /* Ctrl-? */
    term.c_cc[VERASE]   = 0;     /* del */
    term.c_cc[VKILL]    = 0;     /* @ */
    term.c_cc[VEOF]     = 4;     /* Ctrl-d */
    term.c_cc[VTIME]    = 0;
    term.c_cc[VMIN]     = 0;
    term.c_cc[VSWTC]    = 0;     /* '?0' */
    term.c_cc[VSTART]   = 0;     /* Ctrl-q */
    term.c_cc[VSTOP]    = 0;     /* Ctrl-s */
    term.c_cc[VSUSP]    = 0;     /* Ctrl-z */
    term.c_cc[VEOL]     = 0;     /* '?0' */
    term.c_cc[VREPRINT] = 0;     /* Ctrl-r */
    term.c_cc[VDISCARD] = 0;     /* Ctrl-u */
    term.c_cc[VWERASE]  = 0;     /* Ctrl-w */
    term.c_cc[VLNEXT]   = 0;     /* Ctrl-v */
    term.c_cc[VEOL2]    = 0;     /* '?0' */

//  tcflush(fdc, TCIFLUSH);
    n = tcsetattr(fdc, TCSANOW, &term);
over :
    return (n);
    }

bool offsetRequest(std_srvs::Trigger::Request  &req,
         std_srvs::Trigger::Response &res) {
    std::unique_lock<std::mutex> lock(m_);
    offset_reset_ = RESET_COMMAND_TRY;
    while (offset_reset_ > 0) {cv_.wait(lock);}
    lock.unlock();
    res.message = "Reset offset command was send " + std::to_string(RESET_COMMAND_TRY) + " times to the sensor.";
    res.success = true;
    return true;
}

bool clearSocket(const int& fdc, char* leftover){
    int len = 0;
    int c = 0;
    int length = 255;
    while ( len < length ) {
        c = read(fdc, leftover+len, length-len);
        if (c > 0) {
            len += c;
            ROS_DEBUG("More data to clean up; n = %d (%d) ===", c, len);
        } else {
            ROS_DEBUG("No more data on socket");
            break;
        }
    }
    // This could actually check if data was received and may return on timeout with false
    return true;
}

bool readCharFromSocket(const int& fdc, const int& length, char* reply){
    int len = 0;
    int c = 0;
    while ( len < length ) {
        c = read(fdc, reply+len, length-len);
        if (c >= 0) {
            len += c;
        } else {
            ROS_DEBUG("=== need to read more data ... n = %d (%d) ===", c, len);
            continue;
        }
    }
    // This could actually check if data was received and may return on timeout with false
    return true;
}

int main(int argc, char **argv) {
    int fdc;
    int clock = 0;
    double rate;
    std::string devname, frame_id;
    bool auto_adjust = true;
    int frq_div = 1;

    fdc = -1;

    ros::init(argc, argv, "dynpick_driver");
    ros::NodeHandle n, nh("~");
    nh.param<std::string>("device", devname, "/dev/ttyUSB0");
    nh.param<std::string>("frame_id", frame_id, "/sensor");
    nh.param<double>("rate", rate, 1000);
    nh.param<bool>("acquire_calibration", auto_adjust, true);
    nh.param<int>("frequency_div", frq_div, 1);

    ros::ServiceServer service = n.advertiseService("tare", offsetRequest);
    ros::Publisher pub = n.advertise<geometry_msgs::WrenchStamped>("force", 1000);

    ros::AsyncSpinner spinner(2); // Use 2 threads
    spinner.start();

    // Open COM port
    ROS_INFO("Open %s", devname.c_str());

    fdc = open(devname.c_str(), O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (fdc < 0) {
        ROS_ERROR("could not open %s\n", devname.c_str());
        return -1;
    }

    // Obtain sampling rate
    ROS_INFO("Sampling time = %f ms\n", 1.0/rate);

    // Set baud rate of COM port
    SetComAttr(fdc);

    // Clean up
    char trash[255];
    clearSocket(fdc, trash);

    float calib[6] = {1, 1, 1, 1, 1, 1};
    // Autoadjust
    if (auto_adjust){
        write(fdc, "p", 1);
        char reply[CALIB_DATA_LENGTH];
        readCharFromSocket(fdc, CALIB_DATA_LENGTH, reply);
        sscanf(reply,"%f,%f,%f,%f,%f,%f",
        &calib[0], &calib[1], &calib[2], &calib[3], &calib[4], &calib[5]);
        ROS_INFO("Calibration from sensor:\n%.3f LSB/N, %.3f LSB/N, %.3f LSB/N, %.3f LSB/Nm, %.3f LSB/Nm, %.3f LSB/Nm",
        calib[0], calib[1], calib[2], calib[3], calib[4], calib[5]);
        clearSocket(fdc, trash);
    }

    // Set frequncy divider filter
    if (frq_div == 1 || frq_div == 2 || frq_div == 4 || frq_div == 8) {
        char cmd[2];
        sprintf(cmd, "%dF", frq_div);
        write(fdc, cmd, 2);
        ROS_INFO("Set the frequency divider to %s", cmd);

        // check if successful
        write(fdc, "0F", 2);
        char repl[3];
        readCharFromSocket(fdc, 3, repl);
        ROS_ERROR_COND(repl[0]-'0' != frq_div, "Response by sensor is not as expected! Current Filter: %dF", repl[0]-'0');
        clearSocket(fdc, trash);
    } else {
        ROS_WARN("Not setting frequency divider. Parameter out of acceptable values {1,2,4,8}: %d", frq_div);
    }

    // Request for initial single data
    write(fdc, "R", 1);

    ros::Rate loop_rate(rate);
    while (ros::ok()) {
        char str[256];
        int tick;
        unsigned short data[6];

        geometry_msgs::WrenchStamped msg;

        std::unique_lock<std::mutex> lock(m_);
        if (offset_reset_ <= 0){
            // Request for initial data (2nd round)
            write(fdc, "R", 1);

            // Obtain single data
            readCharFromSocket(fdc, DATA_LENGTH, str);

            sscanf(str, "%1d%4hx%4hx%4hx%4hx%4hx%4hx",
                   &tick, &data[0], &data[1], &data[2], &data[3], &data[4], &data[5]);

            msg.header.frame_id = frame_id;
            msg.header.stamp = ros::Time::now();
            msg.header.seq = clock++;

            msg.wrench.force.x = (data[0]-8192)/calib[0];
            msg.wrench.force.y = (data[1]-8192)/calib[1];
            msg.wrench.force.z = (data[2]-8192)/calib[2];
            msg.wrench.torque.x = (data[3]-8192)/calib[3];
            msg.wrench.torque.y = (data[4]-8192)/calib[4];
            msg.wrench.torque.z = (data[5]-8192)/calib[5];

            pub.publish(msg);
        } else {
            // Request for offset reset
            write(fdc, "O", 1);
            offset_reset_ --;
            cv_.notify_all();
        }
        lock.unlock();
        ros::spinOnce();

        loop_rate.sleep();
    }

    return 0;
}
